compiler/rustc_mir/src/const_eval/eval_queries.rs

```diff
@@ -311,7 +311,19 @@ pub fn eval_to_allocation_raw_provider<'tcx>(
             // <https://github.com/rust-lang/rust/issues/71800>.
             let emit_as_lint = if let Some(def) = def.as_local() {
                 // (Associated) consts only emit a lint, since they might be unused.
-                matches!(tcx.def_kind(def.did.to_def_id()), DefKind::Const | DefKind::AssocConst)
+                matches!(tcx.def_kind(def.did.to_def_id()), DefKind::Const | DefKind::AssocConst) || {
```

Comment:
afaict this also only emits this as a lint even if we are using the associated const which doesn't seem desirable to me

what's the result for the following with this change?
```rust
type Foo = [u8; 0 - 1];

fn foo() -> Foo {
   todo!()
}
```
and
```rust
struct Q<const N: usize>;
type Foo = Q<{0 - 1}>;

impl Default for Foo {
    fn default() -> Self {
        Q
    }
}

Reply:
Yes, these will become future incompat lints, too